invoices/invoiceregistry.go

```diff
@@ -997,12 +997,23 @@ func (i *InvoiceRegistry) SubscribeSingleInvoice(
 		}
 	}()
 
+	// Within the lock, we both query the invoice state and pass the client
+	// subscription to the invoiceEvents channel. This is to make sure that
+	// the client receives a consistent stream of events.
+	i.Lock()
+	defer i.Unlock()
+
+	err := i.deliverSingleBacklogEvents(client)
```

Comment:
It seems like this is the only change needed to address the issue? So no need to merge new subscription client events into the `invoiceEvents` channel.

Also, as someone new to this part of the codebase, I find the locking patterns hard to reason about. I noticed there are cases where we acquire the registry's lock while performing database operations and other cases where we don't.

Reply:
I merged the two types of events in one channel to prevent the following scenario:
1. suppose invoice is `open`
2. subscribesingleinvoice is called. the `open` state gets delivered immediately and subscriber is posted to the (separate) new subscribe channel
3. invoice is settled and invoice update is posted to (separate) invoice update channel
4. main loop picks up event from (3) first. subscriber is not yet in the list and doesn't receive a notification of this event
5. main loop picks up new subscriber event from (2) and adds it to the list.
-> event is missed

Reply:
It doesn't seem like step 4 can actually happen if we acquire the lock though? An invoice update requires acquiring the lock, so since it's already acquired by `SubscribeSingleInvoice` when dispatching the new client to the main event loop, this ensures the new client gets handled first before the invoice update can get sent to the main event loop.

Reply:
Yes, the lock is acquired. So `subscribe` first posts to the subscriber channel, releases the lock and then settleinvoice posts to the update channel. If at that moment, the main loop enters the select again and both channels are readable (because it was doing slow processing of a previous event for example), it is undetermined which event is read first. But we do require them to be in the order subscribe then update.

Reply:
> If at that moment, the main loop enters the select again and both channels are readable (because it was doing slow processing of a previous event for example), it is undetermined which event is read first.

Ah I didn't realize the `invoiceEvents` chan is buffered, which usually isn't the norm within event loops in other parts of the codebase.

Reply:
I didn't introduce the buffer, but I think the original reason is to not block the critical payment path when invoices are updated.

Reply:
> I noticed there are cases where we acquire the registry's lock while performing database operations and other cases where we don't.

Do you have a particular location where you saw something you think isn't right?

The general idea is to use the lock to make sure a db update and the event notification are synchronized. If we just read from the database such as in `LookupInvoice`, the lock is not required. The database operations themselves are already atomic because of the bbolt tx.